### ISSUE TITLE
juniper_junos_table fails if subtable contains no elements

### ISSUE THREAD

**Issue by ferfemar** [Juniper/ansible-junos-stdlib#379]
Issue Type
------

 - Bug Report

Module Name
------

juniper_junos_table

Juniper.Junos role and Python libraries version

```
$ ansible --version
ansible 2.5.2
  config file = /home/ferfecky/ansible_playground/detailed-interfaces/ansible.cfg
  configured module search path = [u'/home/ferfecky/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.13 (default, Nov 24 2017, 17:33:09) [GCC 6.3.0 20170516]

$ 

$ pip freeze
ansible==2.5.2
appdirs==1.4.3
asn1crypto==0.24.0
attrs==16.3.0
bcrypt==3.1.4
beautifulsoup4==4.5.3
certifi==2018.4.16
cffi==1.11.5
chainmap==1.0.2
chardet==3.0.4
click==6.6
colorama==0.3.9
configparser==3.5.0
constantly==15.1.0
cryptography==2.2.2
enum34==1.1.6
future==0.16.0
gpg==1.8.0
html5lib==0.999999999
httplib2==0.9.2
icdiff==1.9.1
idna==2.7
incremental==16.10.1
ipaddress==1.0.22
Jinja2==2.10
jsnapy==1.3.1
jtextfsm==0.3.1
junos-eznc==2.1.8
jxmlease==1.0.1
keyring==10.1
keyrings.alt==1.3
librouteros==1.0.4
lxml==4.2.1
M2Crypto==0.24.0
MarkupSafe==1.0
napalm==2.3.1
napalm-ansible==0.9.1
napalm-base==0.25.0
napalm-eos==0.6.1
napalm-fortios==0.4.0
napalm-ios==0.8.1
napalm-iosxr==0.5.6
napalm-junos==0.12.1
napalm-nxos==0.7.1
napalm-panos==0.4.0
napalm-pluribus==0.5.1
napalm-ros==0.2.2
napalm-vyos==0.1.3
ncclient==0.5.3
netaddr==0.7.19
netmiko==2.1.1
netsnmp-python==1.0a1
packaging==16.8
PAM==0.4.2
pan-python==0.12.0
paramiko==2.4.1
pbr==3.1.1
ply==3.10
pyasn1==0.4.3
pyasn1-modules==0.0.7
pycparser==2.18
pycrypto==2.6.1
pycryptodome==3.4.5
pyeapi==0.8.2
pyfg==0.50
pygobject==3.22.0
pyIOSXR==0.53
PyNaCl==1.2.1
pynxos==0.0.3
pyOpenSSL==16.2.0
pyparsing==2.2.0
pyPluribus==0.3.1
pyserial==3.4
pysmi==0.1.2
pysnmp==4.3.2
pysnmp-apps==0.3.2
pysnmp-mibs==0.1.3
python-apt==1.4.0b3
pyxdg==0.25
PyYAML==3.12
requests==2.19.1
requests-toolbelt==0.8.0
rmate==1.0.3
scp==0.11.0
SecretStorage==2.3.1
service-identity==16.0.0
six==1.11.0
stevedore==1.28.0
textfsm==0.4.1
Twisted==16.6.0
typing==3.6.2
urllib3==1.23
virtualenv==15.1.0
virtualenv-clone==0.2.6
VyattaConfParser==0.5.1
webencodings==0.5
xmltodict==0.11.0
zope.interface==4.3.2

```

OS / Environment
------

```
> show version
Hostname: MX-LAB2
Model: mx40-t
Junos: 15.1R6.7
```
Summary
------

Using OP table with subtable (sorry, not the right term I am sure) in juniper_junos_table fails if the subtable contains no elements. Works OK in Junos Pyez outside Ansible.

In particular I am testing it on OP table to retrieve all physical ports and to include also all logical interfaces. When there is an interface without logical-interface element the playbook fails.

Steps to reproduce
------

Playbook:

```yaml
---
- name: test
  hosts: lab
  gather_facts: False
  roles: 
   - Juniper.junos
  tasks:
    - name: Retrieve Interfaces
      juniper_junos_table:
        file: "interface_extensive.yml"
        path: "."
        host: "{{ansible_host}}"
        args: 
      register: response
    - name: Print response
      debug:
        var: response
        verbosity: 2
```
OP table definition:
```
---
LogPortTable:
  rpc: get-interface-information
  args: 
   extensive: True
   interface_name: '[efgx][et]-*'
   # interface_name: 'ge-1/0/0*'
  key: name
  item: physical-interface
  view: PhyPortView

PhyPortView:
  fields:
    int_name: name
    description: description
    admin_status: admin-status
    oper_status: oper-status
    logical_interfaces: _SubIntTable 
    
_SubIntTable: 
  item: logical-interface
  view: _SubIntView
    
_SubIntView:
  fields:
    subint_name: name
    description: description  
```

Expected results

Similar output to this one (which is produced with filter to only include interface with subinterfaces)
```
TASK [Print response] ********************************************************************************************************************************************************************************************************************
task path: /home/ferfecky/ansible_playground/detailed-interfaces/playbook_tables.yml:15
ok: [lab2mx] => {
    "response": {
        "changed": false,
        "failed": false,
        "msg": "Successfully retrieved 1 items from LogPortTable.",
        "resource": [
            {
                "admin_status": "up",
                "description": "MX-LAB3::ge-2/1/8",
                "logical_interfaces": [
                    {
                        "description": "MX-LAB3::ge-2/1/8.100",
                        "subint_name": "ge-1/0/0.100"
                    },
                    {
                        "description": null,
                        "subint_name": "ge-1/0/0.32767"
                    }
                ],
                "name": "ge-1/0/0",
                "oper_status": "down"
            }
        ]
    }
}

```

In Pyez it looks like this: 

```
 ('ge-1/1/6',
  [('admin_status', 'up'),
   ('int_name', 'ge-1/1/6'),
   ('description', None),
   ('oper_status', 'down'),
   ('logical_interfaces', _SubIntTable:172.26.80.44: 0 items)]),
 ('ge-1/1/7',
  [('admin_status', 'up'),
   ('int_name', 'ge-1/1/7'),
   ('description', None),
   ('oper_status', 'down'),
   ('logical_interfaces', _SubIntTable:172.26.80.44: 0 items)]),
```

Actual results
------

```
ansible-playbook 2.5.2
  config file = /home/ferfecky/ansible_playground/detailed-interfaces/ansible.cfg
  configured module search path = [u'/home/ferfecky/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.13 (default, Nov 24 2017, 17:33:09) [GCC 6.3.0 20170516]
Using /home/ferfecky/ansible_playground/detailed-interfaces/ansible.cfg as config file
setting up inventory plugins
Set default localhost to localhost
Parsed /home/ferfecky/ansible_playground/detailed-interfaces/hosts inventory source with ini plugin
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/dist-packages/ansible/plugins/callback/default.pyc
Loading callback plugin jsnapy of type aggregate, v2.0 from /home/ferfecky/.ansible/roles/Juniper.junos/callback_plugins/jsnapy.pyc

PLAYBOOK: playbook_tables.yml ************************************************************************************************************************************************************************************************************
1 plays in playbook_tables.yml

PLAY [test] ******************************************************************************************************************************************************************************************************************************
META: ran handlers
ERROR! Attempting to decrypt but no vault secrets found
ferfecky@script:~/ansible_playground/detailed-interfaces$ ansible-playbook playbook_tables.yml -vvvv --ask-vault-pass
ansible-playbook 2.5.2
  config file = /home/ferfecky/ansible_playground/detailed-interfaces/ansible.cfg
  configured module search path = [u'/home/ferfecky/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.13 (default, Nov 24 2017, 17:33:09) [GCC 6.3.0 20170516]
Using /home/ferfecky/ansible_playground/detailed-interfaces/ansible.cfg as config file
Vault password:
setting up inventory plugins
Set default localhost to localhost
Parsed /home/ferfecky/ansible_playground/detailed-interfaces/hosts inventory source with ini plugin
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/dist-packages/ansible/plugins/callback/default.pyc
Loading callback plugin jsnapy of type aggregate, v2.0 from /home/ferfecky/.ansible/roles/Juniper.junos/callback_plugins/jsnapy.pyc

PLAYBOOK: playbook_tables.yml ************************************************************************************************************************************************************************************************************
1 plays in playbook_tables.yml

PLAY [test] ******************************************************************************************************************************************************************************************************************************
META: ran handlers
Trying secret <ansible.parsing.vault.PromptVaultSecret object at 0x7f810db8f950> for vault_id=default

TASK [Retrieve Interfaces] ***************************************************************************************************************************************************************************************************************
task path: /home/ferfecky/ansible_playground/detailed-interfaces/playbook_tables.yml:10
<172.26.80.44> attempting to start connection
<172.26.80.44> using connection plugin netconf
<172.26.80.44> local domain socket does not exist, starting it
<172.26.80.44> control socket path is /home/ferfecky/.ansible/pc/bf431b6d45
<172.26.80.44> connection to remote device started successfully
<172.26.80.44> local domain socket listeners started successfully
<172.26.80.44>
<172.26.80.44> local domain socket path is /home/ferfecky/.ansible/pc/bf431b6d45
Using module file /home/ferfecky/.ansible/roles/Juniper.junos/library/juniper_junos_table.py
<172.26.80.44> ESTABLISH LOCAL CONNECTION FOR USER: ferfecky
<172.26.80.44> EXEC /bin/sh -c 'echo ~ && sleep 0'
<172.26.80.44> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/ferfecky/.ansible/tmp/ansible-tmp-1529409081.08-56024721213188 `" && echo ansible-tmp-1529409081.08-56024721213188="` echo /home/ferfecky/.ansible/tmp/ansible-tmp-1529409081.08-56024721213188 `" ) && sleep 0'
<172.26.80.44> PUT /home/ferfecky/.ansible/tmp/ansible-local-20127svkGv7/tmpOEbl1b TO /home/ferfecky/.ansible/tmp/ansible-tmp-1529409081.08-56024721213188/juniper_junos_table.py
<172.26.80.44> EXEC /bin/sh -c 'chmod u+x /home/ferfecky/.ansible/tmp/ansible-tmp-1529409081.08-56024721213188/ /home/ferfecky/.ansible/tmp/ansible-tmp-1529409081.08-56024721213188/juniper_junos_table.py && sleep 0'
<172.26.80.44> EXEC /bin/sh -c '/home/ferfecky/Envs/juniper_automation_latest/bin/python /home/ferfecky/.ansible/tmp/ansible-tmp-1529409081.08-56024721213188/juniper_junos_table.py && sleep 0'
<172.26.80.44> EXEC /bin/sh -c 'rm -f -r /home/ferfecky/.ansible/tmp/ansible-tmp-1529409081.08-56024721213188/ > /dev/null 2>&1 && sleep 0'
fatal: [lab2mx]: FAILED! => {
    "changed": false,
    "module_stderr": "No handlers could be found for logger \"ncclient.transport.session\"\nTraceback (most recent call last):\n  File \"/tmp/ansible_7mupc2/ansible_module_juniper_junos_table.py\", line 518, in <module>\n    main()\n  File \"/tmp/ansible_7mupc2/ansible_module_juniper_junos_table.py\", line 514, in main\n    junos_module.exit_json(**results)\n  File \"/home/ferfecky/.ansible/roles/Juniper.junos/module_utils/juniper_junos_common.py\", line 758, in exit_json\n    super(JuniperJunosModule, self).exit_json(**kwargs)\n  File \"/tmp/ansible_7mupc2/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2309, in exit_json\n  File \"/tmp/ansible_7mupc2/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2302, in _return_formatted\n  File \"/tmp/ansible_7mupc2/ansible_modlib.zip/ansible/module_utils/basic.py\", line 517, in remove_values\n  File \"/tmp/ansible_7mupc2/ansible_modlib.zip/ansible/module_utils/basic.py\", line 500, in _remove_values_conditions\nTypeError: Value of unknown type: <class 'jnpr.junos.factory.Table._SubIntTable'>, _SubIntTable:172.26.80.44: 0 items\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 1
}
        to retry, use: --limit @/home/ferfecky/ansible_playground/detailed-interfaces/playbook_tables.retry

PLAY RECAP *******************************************************************************************************************************************************************************************************************************
lab2mx                     : ok=0    changed=0    unreachable=0    failed=1

```

**Comment by kr3ator** [comment]
@vnitinv kindly please assign someone to review the code and accept the patch.

**Comment by kr3ator** [comment]
@vnitinv unfortunatelly this bug has be re-introduced (most probably during migrations to Ansible Collections). It is present for the role branch as well as in the collections.

**Comment by rahkumar651991** [comment]
hi @kr3ator ,

It seems the changes went into the master branch which is going to be used for collections. 
https://github.com/Juniper/ansible-junos-stdlib/blob/master/ansible_collections/Juniper/junos_collection/plugins/modules/juniper_junos_table.py#L330

i will fix this in the roles also. 
https://github.com/Juniper/ansible-junos-stdlib/blame/roles/library/juniper_junos_table.py#L332

**Comment by kr3ator** [comment]
@rahkumar651991 thanks for the commit. Can You please also update the tag as tag 2.4.0 doesn't have that fix and Ansible Galaxy role is pointing to 2.4.0.

**Comment by rahkumar651991** [comment]
@kr3ator we are planning to release new version of roles by next week. Will update once it is done.
